### DISCUSSION TITLE
Custom attribute types and queries on such attributes

### DISCUSSION THREAD

**Discussion by leptonite** [tywalch/electrodb#338] [General]
I'll refer to this [example](https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgYQK4GcYRAQRjKYAI1RgFMAVATzDIBo4BRAOxmBioZbY4ElyQcAL5wAZlGxwA5GQA2ZAMYEIAEyJSAUFoWyAhunQoIs6Ii1w4CiM0xRUS6AAoNFi2FRFZwBXChldKtayVL5kKgBccMyoIERkUHQuru6e3qEBQSEA5n5kzJHRsfGJrnApXj5+GczBcJ6oZAUxcQlJAJRmpaKmjlY28FYmUMjYkMx58BCicADafioMOWR5DPVkALodCEmlwNOOAIQAcs3xAHTA6LysZFnxvcbQI+DWEx0APu+Wj8Ojr6xwAA8cAADHBPt8hs8xhM4AA+OAAJgArMitjtSnAYAALCQAdyiZAJjCgEigjikwFk8iyulkcF0UCyMQmDGANzuUDg7NmILOZxRyPWcDIAA9aEowlI2hiLEIMfK5Vp5RpRKhmEpgNY4LZ2Vk9lQRkMHkNIkboG1IrrmFlOqEYKgoMw4AADJlERwAEgQg2gZ3mQgY3t9UDOSzygbgwZ+ZzWQjaLo0KrVGrY2rAjPQZHN5OtWQASmQwFArQQ9ZajEM7X1MHAQLoYApsXAALxwAD0AD13QAdRyOEHvGYARgAtABOdY9lQAajaDAHQ7Hk+nc4Xg5HE6ns7aPbanvbZzFikcecLxZlFj2cEO9cb2PRpRx+MJxNJTikGagWZF76g0qSRV7UdZ1xgJHNnFKL8s2uGBHDvJsR3WBhhxBecMWgshYPghtEMRZC4FQ9CoMzLDWBw+8ZgAZgIoiSjgGUVQ0GsBh+MIKGxPVW1fJhWHYKhILtCwQFUORIm2TERT4jhIikHN2M4m0pHo0oADd4nQLV8mkYdlNlHV4lU7xGmkXQwDATRSkDDEGzLEhyHQcT9OACIhMxDhaFkvM9Mkiw-AAR1QYB5kiAgGhU1xrMk5hdBAEyJN8jyTKkbyItKAKgpCrE7HofSosxENxI7ds4A4y44DxaAAGt0FWUhuUMdgsWxMhvlJRR4DxXQQiwOBAm5GAAH59IsJKzQwLBcHwQh7MoGgyEBHM4QpXRmCoaU0tcDLgrCUKcs2iw7hgSJMJzA6DOOnUyxtA0zryiL8qvZgVGPRy3IsIgqF4VyEsksAqqc3zXFEYA5FcqQAAUAGkfKB74Xk08hIhmSkVCkZCRuEc70AB97MRBsHZIAZRh86LCsBH2BMmYMaBx7Ioe+j8t+rFdE8ZKAFkqAoNn5Fh6yZQ0JLK2gBSuLbbh+P4MgQEBJKpkhUWVDKm04W0axa3kETtPk5XFNtNsEpc2StesXT6JiuKTZl6xYcKniIMFBgndBdCVRDMWbTOdw4NN5g2jDCBHBldsSrKwxKqgGqGBxVqQwZabiFIVryvQXR1JUBlDCkd1HBdl20KkM51aGT2sjDMg4KQY3pD93ThADrIg5DsPOIj6rauauOfgTuzk4a75mHU2AwixCAGWddlMFWhRWoVnR9EMHNi+YjX4D8S2wgAGRtnW2L18W7RrqQ69hzfre1ir2GbMColisg7Z+AoiRF8l89RBg0MSFVQ7gI4IAJPIGAKVso9XHqgMAKgGytTrnVSYNQQjXljiKUUlw2A2jgH7AedJ0DjyoHIEweIV4ewPl7H2jgN73xUDvbWAc8QtT8P2JA8dIxIDIP5BurYETsJNNABgt9HYfyREItCbRG7Nw0L-cOYhdBUjIIYBsvh1RsCtkweQSgJAkjJJELRphnwAOdJ9OAAARKglsIDGIAEKWC8BMSIAAiAAqjYCBkAR6Z2FhmAwu1ZgQCIAAKw6nAAA8gEjq6wzizH5GcdYJclYq3LuQyhcVqG7zoQwsgTDFZclYSKDh8YuF5N4QkaQud37Ik-m0aU4jg6SNbuVQI8jmBSHgElJsihcY4CZCyAECthZSCQFUQICDQiuUKC0AA3OkYZtRwzaXGfEKZQzMh1FkA0JoRQoBTKEFIAezAIDwEXsALIMV2ZjzKIye+5AuR9PmtIbycT5gJO9qQCheQqE0OsOk+ImTHDMJ7rk9hnCWzcP8sUhgOcsgenKZU3Z+g4DqiqvsvEzp4U5jEYHWpUi24VQ7nAuAvA+q206qtVp4CfyMHUcoKxzVypdSoMXIAA) when describing my issue below.

For an attribute (line 53) of my item I want to use instances of a custom class (line 4). They should be serialized to a string when writing into the DynamoDB table and deserialized into instances of the custom class when reading the table. Furthermore, `EntityItem<...>` should infer the custom class rather than string for this attribute (line 78). This works fine by using `CustomAttributeType`, but my first question is: Did I do it the right way?

Next I want to use this attribute in a query like this: `[...].where(({ myAttribute }, { eq }) => eq(myAttribute, new MyCustomClass([...]))).go()`

I tried this in three ways (lines 100, 103 and 106) and I expected the version in line 100 to work, but it doesn't. I think it could work, ElectroDB just needs to apply the attribute's set function (line 57) not only to attribute values, but also to query values on this attribute.

Do you agree that ElectroDB should work the way I expected?

If not, how should I write queries on such attributes?

**Comment by tywalch** [comment]
Hi @leptonite 👋 

Great question! We have a few options, but I want to first address why filters don't go down the usual path that calls `get`, `set`, and `validate`. Filtering often requires very precise/deliberate choices on behalf of the caller, for example: with the `contains` operation you are likely providing a partial copy of the full attribute, so running `set` or `validate` on partially provided data could have frustrating consequences for users. As a result, the requirements placed on filter values are lean; only validating paths exist and types match, which brings us to your challenge!

There are a number of ways we can address this, but here are the top three stand outs:

## Option 1
*Whenever* you struggle with a filtering issue, remember that the best part of how the `where()` clause works is that it returns the actual string that will be used in the ConditionExpression. There are a few utility operations you can use to help in this case, namely `name()`, and `escape()`. [Here is a playground showing this approach, with a little blurb about what it's doing!](https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgYQK4GcYRAQRjKYAI1RgFMAVATzDIBo4BRAOxmBioZbY4ElyQcAL5wAZlGxwA5GQA2ZAMYEIAEyJSAUFoWyAhunQoIs6Ii1w4CiM0xRUS6AAoNFi2FRFZwBXChldKtayVL5kKgBccMyoIERkUHQuru6e3qEBQSEA5n5kzJHRsfGJrnApXj5+GczBcJ6oZAUxcQlJAJRmpaKmjlY28FYmUMjYkMx58BCicADafioMOWR5DPVkALodCEmlwNOOAIQAcs3xAHTA6LysZFnxvcbQI+DWEx0APu+Wj8Ojr6xwAA8cAADHBPt8hs8xhM4AA+OAAJgArMitjtSnAYAALCQAdyiZAJjCgEigjikwFk8iyulkcF0UCyMQmDGANzuUDg7NmILOZxRyPWcDIAA9aEowlI2hiLEIMfK5Vp5RpRKhmEpgNY4LZ2Vk9lQRkMHkNIkboG1IrrmFlOqEYKgoMw4AADJlERwAEgQg2gZ3mQgY3t9UDOSzygbgwZ+ZzWQjaLo0KrVGrY2rAjPQZHN5OtWQASmQwFArQQ9ZajEM7X1MHAQLoYApsXAALxwAD0AD13QAdRyOEHvGYARgAtABOdY9lQAajaDAHQ7Hk+nc4Xg5HE6ns7aPbanvbZzFikcecLxZlFj2cEO9cb2PRpRx+MJxNJTikGagWZF76g0qSRV7UdZ1xgJHNnFKL8s2uGBHDvJsR3WBhhxBecMWgshYPghtEMRZC4FQ9CoMzLDWBw+8ZgAZgIoiSjgGUVQ0GsBh+MIKGxPVW1fJhWHYKhILtCwQFUORIm2TERT4jhIikHN2M4m0pHo0oADd4nQLV8mkYdlNlHV4lU7xGmkXQwDATRSkDDEGzLEhyHQcT9OACIhMxDhaFkvM9Mkiw-AAR1QYB5kiAgGhU1xrMk5hdBAEyJN8jyTKkbyItKAKgpCrE7HofSosxENxI7ds4A4y44DxaAAGt0FWUhuUMdgsWxMhvlJRR4DxXQQiwOBAm5GAAH59IsJKzQwLBcHwQh7MoGgyEBHM4QpXRmCoaU0tcDLgrCUKcs2iw7hgSJMJzA6DOOnUyxtA0zryiL8qvZgVGPRy3IsIgqF4VyEsksAqqc3zXFEYA5FcqQAAUAGkfKB74Xk08hIhmSkVCkZCRuEc70AB97MRBsHZIAZRh86LCsBH2BMmYMaBx7Ioe+j8t+rFdE8ZKAFkqAoNn5Fh6yZQ0JLK2gBSuLbbh+P4MgQEBJKpkhUWVDKm04W0axa3kETtPk5XFNtNsEpc2StesXT6JiuKTZl6xYcKniIMFBgnaRVF0JVdsSpxVqXUtshHATOAIFoKAGy0hlqQgPFDCoCBUCxCBLCqcgGSYcU-AMLS8Ds0gyCOWLWu6LldDgfV1OdWyZtzjRPean3VLpBoA5dIOQ7D7VHGYCB4AwMI4BavwOjpExo7gWP496hRk9akvGHTshM+sbOq-IAA1RvC9MEuy7yBlpuIava+910F4UMz-cDyuD5Tqf-Ac1O5+LBfNKX-fZvX2QGgq9hsTj+Ar9mgyAwEAFDAHbswGuJVeA6lAGAWot8GytWPmQQKdJ+IinnovZ0IAqBZlkKIA4zE2J6z1GcdwcFTbMDaGcPEA9-aOCQCGLGiAogFwYKfc+wgOgtgRImCw3o-YmgtCINs3oOG0FPNdfUohDQ-E7kSEW5IXYuzQm0No8oXTUKyBAAOGggA)

## Option 2 
The next option involves removing the stringification of your Color class, and filtering on the individual properties. This assumes that the stringified value isn't being used since it's serialized on write and deserialized on retrieval. I assume this option won't work for you, while I am interested in your use-case for serializing the value, but I am sure it's part of a bigger requirement. I mention this as an option though because there is actually a mix of option 1 and option 2 next...

## Option 3
Here is another approach for you to consider: the attribute option watch ([read more here](https://electrodb.dev/en/modeling/attributes/#attribute-watching)) was created to make denormalization easier, and I think it might be helpful here! 

Quick break for story time: one project I had required me to create a "transaction warehouse" that stored user purchase transactions. There was a need to filter by the transaction description (e.g. filter by the term `"starbucks"`). Since DynamoDB is case sensitive, and the descriptions were well-formatted, this posed a problem. This is where `watch` came into play. The watch option allows one attribute to have their `get` and/or `set` attributes triggered when another attribute is triggered. So for the transaction description, I created _another_ attribute called `filterDescription` which watched `description`, and set its value own value to be a lowercase version of the description with no spaces. I also used the `hidden` and `readOnly` options to remove its exposure to the frontend. This allowed for a wildly improved filter experience by users. [Here is an example of that in action](https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgUQHY2DAngGjgZQFMoA3YAY0LgF84AzKCEOAIkIBtDyZGATAIxYBuAFAjyEVAGd4MAIb9OcALyssEAK5QA+vMWFtqOSELCxE6fAAqUOdLndgklXFSEA7inSYsAChFwiAGBcCAQvBwAXEEhIYTe2NEsPHZSDhiSLDjBsSTEUk6oSQCMWTkhUsRklEkp9o6S2u5yUIQAFpqVLOXU2bFyMDzA-BowhFLRCOWBdWkNqACSvJPTIdhghEkyUMCoAOZlsYG9qw4SGuhLK0drWBtbQ-uHN3CtAI4awK3LcDwahKsTjcIlJyDswBkijEXutNqxtrsDn0Xu9Pt9on9CMjYkCjiCwcAIYUAGLAdhjKDXGF3OEsBFPbFHZowchtaIAbRY+PBkJYAF1GbE2sBeBEoZjBSFWnJeAB5VDsLAYqD-SWBSowaK+bR4JDcwmQmgAShUAD5oS84MA6HBfPqic4AITKVQXCJ0XaEXgmqaW2KtGBaVBwe2QgB0MAgABkIO5iABhOSVXxGsOtMDsByEXwAenZAD0ADruPk5vZ4ADkFaNoj9NFWx0BPUZuwiAA9xlTYrN0oUJhajmAANZdl4ejg-FjD56WiTgCAFMYclhnTSXXj8hs0NVwKQjgc3cfsSd7mcvOeQRdwzk9+ZLTeW6iAxlP47IpB6JRPmvmSQyOCrhcMBLC4FYAAwQcUFaiBY-63pCIGqOBkHQeIf7wKGhSgfg8hQCM5BDlIcAALRwAAcnIQ4DEwwBwAACpmABecioSINipL2khhmChADNmCCAeueDwYUSx4JhkjUKmewQCmohsbY9Thh8xBYBGilzJCUi+EggnAbwxrBGG7htMQ2Y6RJqCkuSxC9IgFjyLsUhSWacAOXITl2uMBIOlZZIUpWMgtPhhHVkaRkySmQA)

How that relates to your use case is maybe you can have your `color` attribute, and an `rgb` attribute that watches `color`. [Something like this](https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgYQK4GcYRAQRjKYAI1RgFMAVATzDIBo4BRAOxmBioZbY4ElyQcAL5wAZlGxwA5GQA2ZAMYEIAEyJSAUFoWyAhunQoIs6Ii1w4CiM0xRUS6AAoNFi2FRFZwBXChldKtayVL5kKgBccMyoIERkUHQuru6e3qEBQSEA5n5kzJHRsfGJrnApXj5+GczBcJ6oZAUxcQlJAJRmpaKmjlY28FYmUMjYkMx58BCicADafioMOWR5DPVkALodCEmlwNOOAIQAcs3xAHTA6LysZFnxvcbQI+DWEx0APu+Wj8Ojr6xwAA8cAADHBPt8hs8xhM4AA+OAAJgArMitjtSnAYAALCQAdyiZAJjCgEigjikwFk8iyulkcF0UCyMQmDGANzuUDg7NmILOZxRyPWcDIAA9aEowlI2hiLEIMfK5Vp5RpRKhmEpgNY4LZ2Vk9lQRkMHkNIkboG1IrrmFlOqEYKgoMw4AADJlERwAEgQg2gZ3mQgY3t9UDOSzygbgwZ+ZzWQjaLo0KrVGrY2rAjPQZHN5OtWQASmQwFArQQ9ZajEM7X1MHAQLoYApsXAALxwAD0AD13QAdRyOEHvGYARgAtABOdY9lQAajaDAHQ7Hk+nc4Xg5HE6ns7aPbanvbZzFikcecLxZlFj2cEO9cb2PRpRx+MJxNJTikGagWZF76g0qSRV7UdZ1xgJHNnFKL8s2uGBHDvJsR3WBhhxBecMWgshYPghtEMRZC4FQ9CoMzLDWBw+8ZgAZgIoiSjgGUVRrAYfjCChsT1VtXyYVh2CoSC7QsEBVDkSJtkxEVeI4SIpBzNiOJtKR6NKAA3eJ0C1fJpGHJTZR1eIVO8RppF0MAwE0UpAwxBsyxIch0DEvTgAiQTMQ4WgZLzXSJIsPwAEdUGAeZIgIBplNcKyJOYXQQGM8SfPc4ypC88LSn8wLgqxOx6D0yLMRDMSO3bOB2MuOA8WgABrdBVlIblDHYLFsTIb5SUUeA8V0EIsDgQJuRgAB+PSLESs0MCwXB8EIOzKBoMhARzOEKV0ZgqGlVLXHSoKwhC7KNosO4YEiE1oDNH4OhbBF4p87l9hDR8bs2sgHSdbiIJDf0wgYD61m+mNw2YNoAG5hoi0HfOekC4HVFQyFEdkwhBm6gMsjb3Ucm7RukFLQc6+9IhmKQQykZDQaqFQAHkaioXawtBrMjpvAB9BgkBDYQLoRPMDXe86NpR4RwvZWHRTIBzXIsIgqF4FzrsxMBKoxx74bkFypAABQAaW8x6rBeDTyAJykVBJ-bBfpxWJcxFXZDVgBlbWzYsPXIAN4yZlJ5H+fCvK8rlmBdE8JKAFkqAoQP5B1qyZQ0RLK2geTOLbbg+P4MgQEBRKpkhBOVFKm04W0axa3kYStLkvOFNtNt4ucmTS+sHT6Oi2L6-T6wdYKt6fkcQUGD70F0KY1jK71M53DghvAbDCBHBldtitKwwKqgaqGBxFr2Zs6bSBasr0F0NSVAZQwpHdXvUX7y-B6kM4i6GRObTDZ7HCQOvpCnnSOZnueNAXkqOLLyqjVJqm8fgMimsQXe9VvjMDUrAMIWIIAMmdOyTAK0FAtWzjofQhgcx3w0MxUILcwgABl27lxHvnaudp35SE-jrEhbcy7lXYM2MCUQYpkE7j8AoRJ47kgHgPNCiRh4P1Hk-Cejg-AkJUOQsubQzh4man4fsSB3SC0QCKPyHNWwIjIH5aRWQiAMDPsYi+yIr6WJvm0RRWRZ4yiAA). 

Let me know if this helps, I can go into more detail on any of the above and/or help you brainstorm more ideas based on your use case 👍

**Reply by leptonite** [reply]
Hi @tywalch,

thank you for your detailed response. My actual use case are the classes Instant and LocalDate from js-joda and hopefully in the not too distant future the classes Instant and PlainDate of the Temporal API. But let's stop the discussion about date/time libs before it even starts and go back to the actual topic of this thread ;)

Besides solving this single use case I want to understand the general concept and limitations of custom attribute types in electrodb.

* Option 1: This will work, but it means giving up type safety in queries and it's a bit inconvenient to write.
* Option 2: This may work, but now that you know my use case you probably understand that I prefer stringification.
* Option 3: This means storing the value twice, doesn’t it?

In the meantime I came across another problem with my custom attribute type: I cannot use it as part of an index. That’s an additional requirement which I didn’t address until now. And it’s another reason for stringification and therefore against option 2.

All this makes me think that I am somewhat abusing `CustomAttributeType` and the `get` and `set` callbacks. I tend to turning this discussion into a feature request:

I want to be able to define a custom attribute type that has the following properties:
* It has a string representation and a runtime representation.
* The user provides functions to convert between these representations.
* electrodb uses the string representation to store the value in the database.
* electrodb uses the string representation to build queries for DynamoDB.
* electrodb uses the string representation to concatenate the value into partition or sort keys (also for GSIs).
* electrodb’s API always uses the runtime representation which means the user only works with the runtime representation.
* electrodb uses the provided callbacks for conversion from and to string.

This may look like this:
```ts
import {createCustomStringifyableAttributeType} from electrodb;
// createCustomStringifyableAttributeType<T>(stringify: (runtimeRepr: T) => string, parse: (stringRepr: string) => T) => ???

// create a new attribute type and tell electrodb how to convert it from and to string
const colorAttributeType = createCustomStringifyableAttributeType(stringifyColor, parseColor);

const coloredThing = new Entity(
  {
    model: { /* ... */ },
    attributes: {
      id: {
        type: 'string',
        required: true,
      },
      name: {
        type: 'string',
        required: true,
      },
      color: {
        type: colorAttributeType, // use the new type. get and set callbacks not needed because colorAttributeType knows how to convert
        required: true,
      },
    },
    indexes: {
      byId: { /* ... */ },
      byColor: {
        pk: {
          field: 'PK',
          composite: ['color'], // use custom attribute types in an index, electrodb automatically converts into the string representation
        },
        sk: {
          field: 'SK',
          composite: [],
        },
      },
    },
  }
);

type ColoredThing = EntityItem<typeof coloredThing>

const lemon: ColoredThing = {
  id: 'lemon1',
  name: 'lemon',
  color: new Color(255, 255, 0),
}

coloredThing.put(lemon).go()

const renamedLemon: ColoredThing = {
  id: 'lemon1',
  name: 'lemon with new name',
  color: new Color(255, 255, 0),
}

coloredThing.put(renamedLemon).where(({ color }, { eq }) => eq(color, new Color(255, 255, 0))).go() // use custom attribute type in queries. electrodb knows that literal values which are compared to the color attribute must be converted with stringifyColor
```

I am aware that this feature would be somewhat incomplete:
* It can only convert to and from string, but there may be custom attribute types which are better converted to e. g. binary or object. However, strings are the most flexible because they can be used as part of a composed index attribute. Therefore I think the string version is the most important one.
* There may be use cases for some typesafe prefix quieries. For example lets say we have two custom stringifyable attribute types, one for js-joda LocalDate (2024-03-04) and one for js-joda YearMonth (2024-03). With a LocalDate attribute as part of a sort key we may want to query like `entity.query.byLocalDate({}).begins({localDateAttribute: YearMonth.of(2024, 3)).go()`. Something like this is probably tricky to support.

I'm curious about your opinion on this.
